zcash_client_backend/src/sync.rs

```diff
@@ -0,0 +1,399 @@
+
+                // Delete the now-scanned blocks, because keeping the entire chain
+                // in CompactBlock files on disk is horrendous for the filesystem.
+                block_deletions.push(delete_cached_blocks(fsblockdb_root, block_meta));
+
+                if scan_ranges_updated {
+                    // The suggested scan ranges have been updated, so we re-request.
+                    scan_ranges = db_data.suggest_scan_ranges()?;
+                } else {
+                    // At this point, the cache and scanned data are locally
+                    // consistent (though not necessarily consistent with the
+                    // latest chain tip - this would be discovered the next time
+                    // this codepath is executed after new blocks are received) so
+                    // we can break out of the loop.
+                    break;
+                }
+            }
+            _ => {
+                // Nothing to verify; break out of the loop
+                break;
+            }
+        }
+    }
+
+    // 7) Loop over the remaining suggested scan ranges, retrieving the requested data
+    //    and calling `scan_cached_blocks` on each range.
+    let scan_ranges = db_data.suggest_scan_ranges()?;
+    debug!("Suggested ranges: {:?}", scan_ranges);
+    for scan_range in scan_ranges.into_iter().flat_map(|r| {
+        // Limit the number of blocks we download and scan at any one time.
+        (0..).scan(r, |acc, _| {
+            if acc.is_empty() {
+                None
+            } else if let Some((cur, next)) = acc.split_at(acc.block_range().start + batch_size) {
+                *acc = next;
+                Some(cur)
+            } else {
+                let cur = acc.clone();
+                let end = acc.block_range().end;
+                *acc = ScanRange::from_parts(end..end, acc.priority());
+                Some(cur)
+            }
+        })
+    }) {
+        // Download the blocks in `scan_range` into the block source.
+        let block_meta = download_blocks(client, fsblockdb_root, db_cache, &scan_range).await?;
+
+        // Scan the downloaded blocks.
+        let scan_ranges_updated =
+            scan_blocks(params, fsblockdb_root, db_cache, db_data, &scan_range)?;
+
+        // Delete the now-scanned blocks.
+        block_deletions.push(delete_cached_blocks(fsblockdb_root, block_meta));
+
+        if scan_ranges_updated {
+            // The suggested scan ranges have been updated (either due to a continuity
+            // error or because a higher priority range has been added).
+            info!("Waiting for cached blocks to be deleted...");
+            for deletion in block_deletions {
+                deletion.await?;
+            }
+            return Ok(true);
+        }
+    }
+
+    info!("Waiting for cached blocks to be deleted...");
+    for deletion in block_deletions {
+        deletion.await?;
+    }
```

Comment:
In my perf testing, at least on Windows, the disk I/O associated with storing and soon after deleting the blocks from disk _dominated_ the sync time. By keeping the batch in memory, sync went _much_ faster. And I have a fast SSD by the way. The downside of this approach is with a fixed batch size, the memory requirement for this varies widely, as sandblasted blocks can cause a batch that typically requires small memory to take >2GB of memory. So I have in mind another approach that somewhat does away with batches altogether. We can chat about it if you're interested. Essentially though, it's the idea that we'll ask the server to give us _all_ the blocks we're interested in, and we'll throttle the download speed by reading them no faster than we have space for in a deliberately limited queue that feeds into the scan_blocks fn. This way, we can keep the scan_blocks function busy, memory usage under control, and not worry about some somewhat arbitrary value for 'batch size'.